lib/obs.rb

```diff
@@ -0,0 +1,148 @@
+  # Binary.version => [String]
+  class Binary < Hashie::Mash
+    include Hashie::Extensions::Mash::SymbolizeKeys
+    def self.coerce(binary)
+      if binary.is_a?(Array)
+        binary.map { |bin| Binary.new(bin) }
+      else
+        [Binary.new(binary)]
+      end
+    end
+  end
+
+  # A collection of one more binaries with the following properties
+  # Collection.binaries => [Array] of [Binary]
+  # Collection.matches  => [Integer]
+  class Collection < Hashie::Mash
+    include Hashie::Extensions::Mash::SymbolizeKeys
+    include Hashie::Extensions::Coercion
+    coerce_key :binary, Binary
+    coerce_key :matches, Integer
+
+    def self.coerce(hash)
+      c = Collection.new(hash)
+      c[:binaries] = c.delete(:binary)
+      c
+    end
+  end
+
+  # Helper class that turns a "collection" attribute in HTTP responses into an
+  # instance of [Collection]
+  class Response < Hashie::Mash
+    include Hashie::Extensions::Mash::SymbolizeKeys
+    include Hashie::Extensions::Coercion
+    coerce_key :collection, Collection
+  end
+
+  # Utility function to generate a xpath query from a search term and the following options:
+  # @param [Hash] opts the options for the query.
+  # @option opts [String] :baseproject OBS base project used to search packages for
+  # @option opts [String] :project OBS specific project used to search packages into
+  # @option opts [Boolean] :exclude_debug Exclude Debug packages from search
+  # @option opts [String] :exclude_filter Exclude packages containing this term
+  #
+  def self.xpath_for(query, opts = {})
+    baseproject = opts[:baseproject]
+    project = opts[:project]
+    exclude_debug = opts[:exclude_debug]
+    exclude_filter = opts[:exclude_filter]
+
+    words = query.split(" ").reject { |part| part.match(/^[0-9_\.-]+$/) }
+    versrel = query.split(" ").select { |part| part.match(/^[0-9_\.-]+$/) }
+    Rails.logger.debug "splitted words and versrel: #{words.inspect} #{versrel.inspect}"
+    raise InvalidSearchTerm, "Please provide a valid search term" if words.blank? && versrel.blank?
+    raise InvalidSearchTerm, "The package name is required when searching for a version" if words.blank? && versrel.present?
+
+    xpath_items = []
+    xpath_items << "@project = '#{project}' " unless project.blank?
+    substring_words = words.reject { |word| word.match(/^".+"$/) }.map { |word| "'#{word.gsub(/['"()]/, '')}'" }.join(", ")
+    xpath_items << "contains-ic(@name, " + substring_words + ")" unless substring_words.blank?
+    words.select { |word| word.match(/^".+"$/) }.map { |word| word.delete("\"") }.each do |word|
+      xpath_items << "@name = '#{word.gsub(/['"()]/, '')}' "
+    end
+    xpath_items << "path/project='#{baseproject}'" unless baseproject.blank?
+    xpath_items << "not(contains-ic(@project, '#{exclude_filter}'))" if !exclude_filter.blank? && project.blank?
+    xpath_items << versrel.map { |part| "starts-with(@versrel,'#{part}')" }.join(" and ") unless versrel.blank?
```

Comment:
`join(" and ")` this will add many `and`s if there are more than two elements. Maybe you want to use `to_sentence` instead